slay/framework.py

```diff
@@ -0,0 +1,618 @@
+
+_SIMPLE_TYPES = {int, float, complex, bool, str, bytes, None}
+_SIMPLE_CONTAINERS = {list, dict}
+
+
+# Checking of processor class definition ###############################################
+
+
+def _validate_io_type(param: inspect.Parameter) -> None:
+    """
+    For processor I/O (both data or parameters) we allow simple types
+    (int, str, float...) and `list` or `dict` containers of these.
+    Any deeper nested and structured data must be typed as a pydnatic model.
+    """
+    anno = param.annotation
+    if anno in _SIMPLE_TYPES:
+        return
+    if isinstance(anno, types.GenericAlias):
+        if get_origin(anno) not in _SIMPLE_CONTAINERS:
+            raise definitions.APIDefinitonError(
+                f"For generic types, only containers {_SIMPLE_CONTAINERS} are "
+                f"allowed, but got `{param}`."
+            )
+        args = get_args(anno)
+        for arg in args:
+            if arg not in _SIMPLE_TYPES:
+                raise definitions.APIDefinitonError(
+                    f"For generic types, only arg types {_SIMPLE_TYPES} are "
+                    f"allowed, but got `{param}`."
+                )
+        return
+    if issubclass(anno, pydantic.BaseModel):
+        try:
+            anno.schema()
+        except Exception as e:
+            raise definitions.APIDefinitonError(
+                "Pydantic annotations must be able to generate a schema. "
+                f"Please fix `{param}`."
+            ) from e
+        return
+
+    raise definitions.APIDefinitonError(anno)
+
+
+def _validate_endpoint_params(
+    params: list[inspect.Parameter], cls_name: str
+) -> list[tuple[str, definitions.TypeDescriptor]]:
+    if len(params) == 0:
+        raise definitions.APIDefinitonError(
+            f"`{cls_name}.{definitions.ENDPOINT_NAME}` must be a method, i.e. "
+            "with `self` argument."
+        )
+    if params[0].name != definitions.SELF_ARG_NAME:
+        raise definitions.APIDefinitonError(
+            f"`{cls_name}.{definitions.ENDPOINT_NAME}` must be a method, i.e. "
+            "with `self` argument."
+        )
+    input_name_and_types = []
+    for param in params[1:]:  # Skip self argument.
+        if param.annotation == inspect.Parameter.empty:
+            raise definitions.APIDefinitonError(
+                "Inputs of endpoints must have type annotations. "
+                f"For `{cls_name}` got:\n{param}"
+            )
+        _validate_io_type(param)
+        type_descriptor = definitions.TypeDescriptor(raw=param.annotation)
+        input_name_and_types.append((param.name, type_descriptor))
+    return input_name_and_types
+
+
+def _validate_and_describe_endpoint(
+    cls: Type[definitions.ABCProcessor],
+) -> definitions.EndpointAPIDescriptor:
+    """The "endpoint method" of a processor must have the follwing signature:
+
+    ```
+    [async] def run(
+        self, [param_0: anno_0, param_1: anno_1 = default_1, ...]) -> ret_anno:
```

Comment:
The generated stubs make this transparent take care of serialization and deserialization internally. E.g. 

```
class SplitText(stub.StubBase):
    def __init__(self, url: str, api_key: str) -> None:
        self._remote = stub.BasetenSession(url, api_key)

    async def run(self, data: str, num_partitions: int) -> tuple[list, int]:
        json_args = {"data": data, "num_partitions": num_partitions}
        json_result = await self._remote.predict_async(json_args)
        return (json_result[0], json_result[1])
```

Note that I already have a TODO to introduce helper functions for this instead of generating above code verbatim.